File: examples/java/packages/data/JsonFileStore.java
package packages.data;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.Objects;
import java.util.UUID;

public class JsonFileStore<T> {
  private static final ObjectMapper MAPPER = new ObjectMapper();

  @FunctionalInterface
  public interface CheckedUpdater<T> {
    T apply(T current) throws Exception;
  }

  private final Path filePath;
  private final Path lockFilePath;
  private final TypeReference<T> typeReference;
  private final T fallbackData;

  public JsonFileStore(Path filePath, TypeReference<T> typeReference, T fallbackData) {
    this.filePath = filePath;
    this.lockFilePath = filePath.resolveSibling(filePath.getFileName() + ".lock");
    this.typeReference = typeReference;
    this.fallbackData = fallbackData;
  }

  public Path filePath() {
    return filePath;
  }

  public synchronized T read() throws Exception {
    return withFileLock(() -> {
      ensureFileExists();
      return readJsonFile();
    });
  }

  public synchronized void write(T value) throws Exception {
    withFileLock(() -> {
      ensureFileExists();
      writeJsonAtomically(value);
      return null;
    });
  }

  public synchronized void reset() throws Exception {
    withFileLock(() -> {
      ensureFileExists();
      writeJsonAtomically(fallbackData);
      return null;
    });
  }

  public synchronized T update(CheckedUpdater<T> updater) throws Exception {
    return withFileLock(() -> {
      ensureFileExists();
      T current = readJsonFile();
      T next = updater.apply(current);
      if (!Objects.equals(next, current)) {
        writeJsonAtomically(next);
      }
      return next;
    });
  }

  private interface CheckedSupplier<T> {
    T get() throws Exception;
  }

  private <R> R withFileLock(CheckedSupplier<R> supplier) throws Exception {
    Files.createDirectories(lockFilePath.getParent());

    try (FileChannel channel = FileChannel.open(
        lockFilePath,
        StandardOpenOption.CREATE,
        StandardOpenOption.WRITE
    );
         FileLock ignored = channel.lock()) {
      return supplier.get();
    }
  }

  private void ensureFileExists() throws IOException {
    Files.createDirectories(filePath.getParent());
    if (!Files.exists(filePath)) {
      writeJsonAtomically(fallbackData);
    }
  }

  private T readJsonFile() throws IOException {
    return MAPPER.readValue(Files.readString(filePath), typeReference);
  }

  private void writeJsonAtomically(T value) throws IOException {
    Path tempFile = filePath.resolveSibling(
        filePath.getFileName() + "." + ProcessHandle.current().pid() + "." + UUID.randomUUID() + ".tmp"
    );

    Files.createDirectories(filePath.getParent());
    Files.writeString(tempFile, MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(value) + "\n");

    try {
      Files.move(
          tempFile,
          filePath,
          StandardCopyOption.REPLACE_EXISTING,
          StandardCopyOption.ATOMIC_MOVE
      );
    } catch (AtomicMoveNotSupportedException ignored) {
      Files.move(tempFile, filePath, StandardCopyOption.REPLACE_EXISTING);
    } finally {
      Files.deleteIfExists(tempFile);
    }
  }
}

File: examples/kotlin/packages/data/json_file_store.kt
package app.protocol.examples.kotlin.packages.data

import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json

interface JsonFileStore<T> {
    val filePath: String

    suspend fun read(): T

    suspend fun write(value: T)

    suspend fun reset()

    suspend fun update(updater: suspend (current: T) -> T): T
}

data class JsonFileStoreOptions<T>(
    val filePath: String,
    val fallbackData: T,
    val serializer: KSerializer<T>,
)

private val json =
    Json {
        prettyPrint = true
        encodeDefaults = true
        ignoreUnknownKeys = false
    }

fun <T> createJsonFileStore(
    options: JsonFileStoreOptions<T>,
): JsonFileStore<T> {
    val mutex = Mutex()
    val path = Paths.get(options.filePath)
    val lockPath = Paths.get("${options.filePath}.lock")

    suspend fun ensureFileExists() {
        withContext(Dispatchers.IO) {
            Files.createDirectories(path.parent)
            if (!Files.exists(path)) {
                writeJsonAtomically(path, options.fallbackData, options.serializer)
            }
        }
    }

    suspend fun <TResult> withFileLock(block: suspend () -> TResult): TResult =
        withContext(Dispatchers.IO) {
            Files.createDirectories(lockPath.parent)
            FileChannel.open(
                lockPath,
                StandardOpenOption.CREATE,
                StandardOpenOption.WRITE,
            ).use { channel ->
                channel.lock().use {
                    block()
                }
            }
        }

    suspend fun readJsonFile(): T =
        withContext(Dispatchers.IO) {
            val content = Files.readString(path)
            json.decodeFromString(options.serializer, content)
        }

    return object : JsonFileStore<T> {
        override val filePath: String = options.filePath

        override suspend fun read(): T =
            mutex.withLock {
                withFileLock {
                    ensureFileExists()
                    readJsonFile()
                }
            }

        override suspend fun write(value: T) {
            mutex.withLock {
                withFileLock {
                    ensureFileExists()
                    writeJsonAtomically(path, value, options.serializer)
                }
            }
        }

        override suspend fun reset() {
            mutex.withLock {
                withFileLock {
                    ensureFileExists()
                    writeJsonAtomically(path, options.fallbackData, options.serializer)
                }
            }
        }

        override suspend fun update(updater: suspend (current: T) -> T): T =
            mutex.withLock {
                withFileLock {
                    ensureFileExists()
                    val current = readJsonFile()
                    val next = updater(current)
                    if (next != current) {
                        writeJsonAtomically(path, next, options.serializer)
                    }
                    next
                }
            }
    }
}

private suspend fun <T> writeJsonAtomically(
    path: Path,
    value: T,
    serializer: KSerializer<T>,
) {
    withContext(Dispatchers.IO) {
        Files.createDirectories(path.parent)
        val tempFile = Files.createTempFile(path.parent, path.fileName.toString(), ".tmp")
        try {
            val content = "${json.encodeToString(serializer, value)}\n"
            Files.writeString(tempFile, content, StandardCharsets.UTF_8)
            Files.move(
                tempFile,
                path,
                java.nio.file.StandardCopyOption.ATOMIC_MOVE,
                java.nio.file.StandardCopyOption.REPLACE_EXISTING,
            )
        } finally {
            Files.deleteIfExists(tempFile)
        }
    }
}
